core/src/main/java/com/distkv/core/block/BlockPool.java

```diff
@@ -0,0 +1,126 @@
+package com.distkv.core.block;
+
+import sun.misc.Unsafe;
+
+import java.io.Closeable;
+import java.lang.reflect.Field;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class BlockPool implements Closeable {
+
+  private static Unsafe getUnsafe() {
+    try {
+      Field field = Unsafe.class.getDeclaredField("theUnsafe");
+      field.setAccessible(true);
+      return (Unsafe) field.get(null);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // the number of block size is a multiple of page size.
+  private static final int BLOCK_SIZE = getUnsafe().pageSize() << 8;
+  private static final int INIT_BLOCK_NUMBER = 1 << 4; // 16
+  private static final int BLOCK_NUMBER_UNIT = INIT_BLOCK_NUMBER; // 16
+  private static final int LOW_WATER_MARK = INIT_BLOCK_NUMBER >> 1; // 8
+  private static final int HIGH_WATER_MARK = INIT_BLOCK_NUMBER << 1; // 32
+
+  private static BlockPool instance;
+  private final int blockSize;
+  private final ConcurrentLinkedQueue<Block> blocks = new ConcurrentLinkedQueue<>();
+
+  private AtomicBoolean isAllocating = new AtomicBoolean(false);
+  private AtomicBoolean isRecycling = new AtomicBoolean(false);
+
+  public static BlockPool getInstance() {
+    return getInstance(BLOCK_SIZE, INIT_BLOCK_NUMBER);
```

Comment:
I'm thinking that it's not a good idea to make `BlockPool` as a singleton.

How about making one shard contains a `BlockPool`? So that we needn't use any `Concurrent` collections and `atomic` member variables in this `BlockPool` class. Because we have no any race conditions between shards.

How do you think of this?

Reply:
One problem is memory utilization because Every `BlockPool` instance will pre allocate some block. if there are too much shards, the memory utilization will be low. BlockPool need to publish memory metric for monitor later. Singleton will make it's easier.

Because the segment shoudn't get blocks from blockpool too frequently. it's will mitigate performance issues.

Reply:
I see. `Singleton` sounds reasonable now.